server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -313,6 +308,275 @@ public TimeValue timeout() {
         }, "create_snapshot [" + snapshotName + ']', listener::onFailure);
     }
 
+    private static void ensureSnapshotNameNotRunning(List<SnapshotsInProgress.Entry> runningSnapshots, String repositoryName,
+                                                     String snapshotName) {
+        if (runningSnapshots.stream().anyMatch(s -> {
+            final Snapshot running = s.snapshot();
+            return running.getRepository().equals(repositoryName) && running.getSnapshotId().getName().equals(snapshotName);
+        })) {
+            throw new InvalidSnapshotNameException(repositoryName, snapshotName, "snapshot with the same name is already in-progress");
+        }
+    }
+
+    private static Map<String, IndexId> getInFlightIndexIds(List<SnapshotsInProgress.Entry> runningSnapshots, String repositoryName) {
+        return runningSnapshots.stream().filter(entry -> entry.repository().equals(repositoryName))
+                .flatMap(entry -> entry.indices().stream()).distinct()
+                .collect(Collectors.toMap(IndexId::getName, Function.identity()));
+    }
+
+    public void cloneSnapshot(CloneSnapshotRequest request, ActionListener<Void> listener) {
+        final String repositoryName = request.repository();
+        final String snapshotName = indexNameExpressionResolver.resolveDateMathExpression(request.target());
+        validate(repositoryName, snapshotName);
+        Repository repository = repositoriesService.repository(repositoryName);
+        if (repository.isReadOnly()) {
```

Comment:
Maybe move this few steps up, after `final String repositoryName = request.repository();`

Reply:
done :)